app/worker.go

```diff
@@ -62,30 +62,37 @@ func (w AppWorker) Start() {
 
 // HandleTransaction handles new instance of TransactionWork.
 func (w *AppWorker) HandleTransaction(work TransactionWork) {
-	ok := tpool.Set(work.Transaction, chain)
+	validTransaction := tpool.Set(work.Transaction, chain)
 
 	// Respond to the request if a response method was provided.
 	if work.Responder != nil {
 		work.Responder.Lock()
 		defer work.Responder.Unlock()
-		work.Responder.Send(ok)
+		work.Responder.Send(validTransaction)
 	}
 }
 
-// HandleBlock handles TransactionWork.
+// HandleBlock handles new instance of BlockWork.
 func (w *AppWorker) HandleBlock(work BlockWork) {
-	ok, _ := chain.ValidBlock(work.Block)
-	if ok {
+	validBlock := tpool.Update(work.Block, chain)
+
+	if validBlock {
+		user := GetCurrentUser()
+		// Append to the chain before requesting
+		// the next block so that the block
+		// numbers make sense.
 		chain.AppendBlock(work.Block)
+		address := user.Wallet.Public()
+		blk := tpool.NextBlock(chain, address, user.BlockSize)
 		if miner.IsMining() {
```

Comment:
I think we should check if the transactions in the new block collide with the block we are currently hashing. Only if there are collisions we should restart the miner. Otherwise restarting it would be a waste.